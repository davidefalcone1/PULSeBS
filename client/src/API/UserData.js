class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

    static fromJson(json){
        const temp =  Object.assign(new UserData(), json);
        return temp;
    }
}

export default UserData;