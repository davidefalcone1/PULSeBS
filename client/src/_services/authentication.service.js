import { BehaviorSubject } from 'rxjs';


const currentUserSubject = new BehaviorSubject(JSON.parse(localStorage.getItem('user')));

export const authenticationService = {
    login,
    logout,
    user: currentUserSubject.asObservable(),
    get currentUserValue () { return currentUserSubject.value }
};

function handleResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if ([401, 403].indexOf(response.status) !== -1) {
                // auto logout if 401 Unauthorized or 403 Forbidden response returned from api
                authenticationService.logout();
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

function login(username, password) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ username, password })
    };
    return new Promise(async function (resolve, reject) {
        try {
            const response = await fetch('/users/authenticate', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ username, password }),
            });
            const user = await response.json().then(handleResponse);
            if (response.ok) {
                localStorage.setItem('user', JSON.stringify(user));
                currentUserSubject.next(user);
                resolve(user);
            }
            else
                reject();
        }
        catch (e) {
            reject();
        }
    });

}
function logout() {

    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
    };
    localStorage.removeItem('user');
    currentUserSubject.next(null);
    return fetch('/logout',requestOptions);
}