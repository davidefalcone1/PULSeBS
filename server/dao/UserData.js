class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

}

module.exports = UserData;