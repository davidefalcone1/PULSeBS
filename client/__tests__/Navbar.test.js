import React from 'react';
import renderer from 'react-test-renderer';
import { AuthContext } from '../src/_services/AuthContext';
import NavBar from '../src/components/Navbar';
import { BrowserRouter } from 'react-router-dom';
import UserData from '../src/API/UserData';

describe('Navbar', ()=>{
    test('Navbar renders correctly when user is a student', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{isStudent: true, user: new UserData(1, '12345', 'Davide Falcone', 'davide.falcone@polito.it')}}>
                    <NavBar/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('Navbar renders correctly when user is a teacher', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{isTeacher: true, user: new UserData(1, '123456', 'Mario Rossi', 'mario.rossi@polito.it')}}>
                    <NavBar/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('Navbar renders correctly when user is the booking manager', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{isBookingManager: true, user: new UserData(1, '123456', 'Mario Rossi', 'mario.rossi@polito.it')}}>
                    <NavBar/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('Navbar renders correctly when user is the support officier', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{isSupportOfficier: true, user: new UserData(1, '123456', 'Mario Rossi', 'mario.rossi@polito.it')}}>
                    <NavBar/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});