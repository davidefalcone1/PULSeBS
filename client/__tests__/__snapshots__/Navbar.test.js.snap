// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Navbar renders correctly when user is a student 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark"
>
  <button
    aria-controls="sidebar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    data-target="#sidebar"
    data-toggle="collapse"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <span
    className="navbar-brand"
  >
    PULSeBS Project
  </span>
  <div
    className="mr-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <span
      className="navbar-brand"
    >
      | Tap on the image to go to the next tutorial image! |
    </span>
  </div>
  <div
    className="ml-md-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <span
      className="navbar-brand"
    >
      Welcome 
      !
    </span>
    <a
      className="nav-link"
      data-rb-event-key={null}
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      Logout
    </a>
  </div>
</nav>
`;

exports[`Navbar Navbar renders correctly when user is a teacher 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark"
>
  <button
    aria-controls="sidebar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    data-target="#sidebar"
    data-toggle="collapse"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <span
    className="navbar-brand"
  >
    PULSeBS Project
  </span>
  <div
    className="mr-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/myCoursesLessonslist"
      onClick={[Function]}
    >
       My courses details
    </a>
  </div>
  <div
    className="ml-md-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <span
      className="navbar-brand"
    >
      Welcome 
      !
    </span>
    <a
      className="nav-link"
      data-rb-event-key={null}
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      Logout
    </a>
  </div>
</nav>
`;

exports[`Navbar Navbar renders correctly when user is the booking manager 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark"
>
  <button
    aria-controls="sidebar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    data-target="#sidebar"
    data-toggle="collapse"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <span
    className="navbar-brand"
  >
    PULSeBS Project
  </span>
  <div
    className="mr-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/monitorUsage"
      onClick={[Function]}
    >
       Monitor usage of the system
    </a>
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/generateContactTracing"
      onClick={[Function]}
    >
       Generate contact tracing
    </a>
  </div>
  <div
    className="ml-md-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <span
      className="navbar-brand"
    >
      Welcome 
      !
    </span>
    <a
      className="nav-link"
      data-rb-event-key={null}
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      Logout
    </a>
  </div>
</nav>
`;

exports[`Navbar Navbar renders correctly when user is the support officier 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark"
>
  <button
    aria-controls="sidebar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    data-target="#sidebar"
    data-toggle="collapse"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <span
    className="navbar-brand"
  >
    PULSeBS Project
  </span>
  <div
    className="mr-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/configureStudentsList"
      onClick={[Function]}
    >
       Students
    </a>
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/configureCoursesList"
      onClick={[Function]}
    >
       Courses
    </a>
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/configureTeachersList"
      onClick={[Function]}
    >
       Teachers
    </a>
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/configureLessonsList"
      onClick={[Function]}
    >
       Lessons
    </a>
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/configureClassesList"
      onClick={[Function]}
    >
       Classes
    </a>
    <a
      aria-current={null}
      className="nav-link"
      data-rb-event-key={null}
      disabled={false}
      href="/configureClassroomsList"
      onClick={[Function]}
    >
       Classrooms
    </a>
  </div>
  <div
    className="ml-md-auto navbar-nav"
    onKeyDown={[Function]}
  >
    <span
      className="navbar-brand"
    >
      Welcome 
      !
    </span>
    <a
      className="nav-link"
      data-rb-event-key={null}
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      Logout
    </a>
  </div>
</nav>
`;
